README.md
# No More 404s [![Build Status](https://travis-ci.org/internetarchive/FirefoxNoMore404s.svg?branch=master)](https://travis-ci.org/internetarchive/FirefoxNoMore404s) [![devDependencies Status](https://david-dm.org/internetarchive/FirefoxNoMore404s/dev-status.svg)](https://david-dm.org/internetarchive/FirefoxNoMore404s?type=dev)

## Developing

This is a "WebExtension". See https://developer.mozilla.org/Add-ons/WebExtensions/Your_first_WebExtension

Develop through the web-ext run command.
```
npm run dev
```


## How to load a zip or xpi file directly into Firefox

- Upgrade to Firefox 45 or newer
- Visit `about:debugging`
- Click "Load Temporary Add-on"
- Click "manifest.json"

If something is wrong, it will silently fail. Try removing the add-on, waiting a few seconds, then re-adding.

More info:
https://developer.mozilla.org/docs/Tools/about%3Adebugging#Enabling_add-on_debugging

Also:

1. Open new tab.
2. Type "about:config".
3. Enter "xpinstall.signatures.required" in search bar.
4. Modify value to "false" (just click on it).
5. Restart Firefox and try install extension again.


## Publishing & Signing

The "web-ext" has a command that publishes and signs the extension at the same time. As far as I can tell, there is no way to do these separately. Once a version is published, it cannot be undone, so use with caution.

First update the version (see Places to Update the version).

```
# This will sign, publish, and update the ./build directory
npm install
npm run sign
```

## Building an unsigned extension

Only use this if you need to produce an unsigned xpi. Use "run" for development
```
npm run build_unsigned
```

## Places to update the version

- Changelog.md - update change log
- src/manifest.json - update version
- src/scripts/background.js - update header
- package.json - update header

## License

Copyright Internet Archive, 2016
AGPL-3


## Credits

  - Richard Caceres, @rchrd2
  - Greg Lindahl, @wumpus
  - Blake Winton, @bwinton, author of https://github.com/bwinton/404archive
  - Adam Miller, @adam-miller, author of https://github.com/adam-miller/ChromeNoMore404s


Changelog.md
## v1.8.2
- Bug fixes #82, #83. Handled DNS error, renamed car.gif to logo.gif across all files, minor UI change

## v1.8.1
- Removed redundant code, added feature to check availability of the URL in archive for First Version & Recent Version

## v1.8.0
- Add toolbar button for Save Page Now, Recent Version, First Version

## v1.7.0
- Major design change by @johngruen. No longer a banner, but rather a modal.

## v1.6.0
- Support additional easy to detect HTTP failures.
- Applied lossless zopfli compression to PNG icons.
- point updateurl to the main site.

## v1.5.6
- Make animation a bit more aggressive by @johngruen

## v1.5.5
- Update exclusion policy and add localhost domains
- Raise strict_min_version to FF 48
- More CSS fixes

## v1.5.4
- Remove code injection from background.js in favor of message passing
- Add validation of response URL
- Fix ES Lint issue
- Update metrics documentation

## v1.5.3
- Fix issue with 404 pages on Myspace.com
- Remove some unused code.
- Move `version` into Telemetry payload

## v1.5.2
- Version bump to test automatic updating

## v1.5.1
- Add external update URL

## v1.5.0
- Add more resilient CSS
- Add test pilot telemetry

## v1.4.6
- Change the text in the banner.

## v1.4.5
- Update the build tooling to use NPM package.json (@pdehaan)
- Update the icons (@johngruen)
- Autofocus the link in the dropdown for accessibility

## v1.4.4
- Sign the extension

## v1.4.3
- Remove background:true from manifest which broke plugin in new FF version.

## v1.4.2
- Update the API call to use POST

## v1.4.1
- Add another header to api call
- Rename the built xpi file

## v1.4.0
- Add special header for endpoint.
- Rename to Wayback Machine.

## v1.3.2
- Use SVG icon
- Add border-box to button element. Fixes rendering issue on some pages.
- Change font to "14px message-box"

## v1.3.1
- Add missing space between text and link and correct typo.

## v1.3.0
- Add ES Lint file and update code style
- Fix icon image path
- Update styles based on suggestions by @johngruen

## v1.2.0
- Disable extension in incognito mode.

## v1.1.2
- Change client.js to use DOM methods instead of setting innerHTML for security.

## v1.1.1
- Remove custom headers and update User Agent to have plugin version.

## v1.1.0
- Add custom User Agent to requests
- Update requests to use https


src/scripts/popup.js
function save_now_function(){
	var wb_url = "https://web.archive.org/save/";
	chrome.runtime.sendMessage({message: "openurl", wayback_url: wb_url, method:'save' }).then(handleResponse, handleError);;	
}

function recent_capture_function(){
	var wb_url = "https://web.archive.org/web/2/";
	chrome.runtime.sendMessage({message: "openurl", wayback_url: wb_url, method:'recent' }, function(response) {
	});
}

function first_capture_function(){
	var wb_url = "https://web.archive.org/web/0/";
	chrome.runtime.sendMessage({message: "openurl", wayback_url: wb_url, method:'first' }, function(response) {
	});
}

function view_all_function(){
	var pattern = /https:\/\/web\.archive\.org\/web\/(.+?)\//g;
	url = document.location.href.replace(pattern, "");
	open_url = "https://web.archive.org/web/*/"+encodeURI(url);
	document.location.href = open_url;
}


document.getElementById('save_now').onclick = save_now_function;
document.getElementById('recent_capture').onclick = recent_capture_function;
document.getElementById('first_capture').onclick = first_capture_function;

chrome.runtime.onMessage.addListener(function(message,sender,sendResponse){
  if(message.message=='urlnotfound'){
  	alert("URL not found in wayback archives!");
  }
});

src/scripts/client.js
/*
 * LICENSE: AGPL-3
 * Copyright 2016, Internet Archive
 */
(function() {
  var enforceBannerInterval;
  var archiveLinkWasClicked = false;
  var bannerWasShown = false;
  var bannerWasClosed = false;

  /**
   * Brute force inline css style reset
   */
  function resetStyesInline(el) {
    el.style.margin = 0;
    el.style.padding = 0;
    el.style.border = 0;
    el.style.fontSize = "100%";
    el.style.font = "inherit";
    el.style.fontFamily = "sans-serif";
    el.style.verticalAlign = "baseline";
    el.style.lineHeight = "1";
    el.style.boxSizing = "content-box";
    el.style.overflow = "unset";
    el.style.fontWeight = "inherit";
    el.style.height = "auto";
    el.style.position = "relative";
    el.style.width = "auto";
    el.style.display = "inline";
    el.style.backgroundColor = "transparent";
    el.style.color = "#333";
    el.style.textAlign = "left";
  }

  /**
   * Communicates with background.js
   * @param action {string}
   * @param complete {function}
   */
  function sendTelemetry(action, complete) {
    chrome.runtime.sendMessage({action: action}, complete);
  }

  /**
   * @param {string} type
   * @param {function} handler(el)
   * @param remaining args are children
   * @returns {object} DOM element
   */
  function createEl(type, handler) {
    var el = document.createElement(type);
    resetStyesInline(el);
    if (handler !== undefined) {
      handler(el);
    }
    // Append *args to created el
    for (var i = 2; i < arguments.length; i++) {
      el.appendChild(arguments[i]);
    }
    return el;
  }

  function createBanner(wayback_url) {
    if (document.getElementById("no-more-404s-message") !== null) {
      return;
    }
    document.body.appendChild(
      createEl("div",
        function(el) {
          el.id = "no-more-404s-message";
          el.style.background = "rgba(0,0,0,.6)";
          el.style.position = "fixed";
          el.style.top = "0";
          el.style.right = "0";
          el.style.bottom = "0";
          el.style.left = "0";
          el.style.zIndex = "999999999";
          el.style.display = "flex";
          el.style.alignItems = "center";
          el.style.justifyContent ="center";
          },
          createEl("div",
            function(el) {
              el.id = "no-more-404s-message-inner";
              el.style.flex = "0 0 420px";
              el.style.position = "relative";
              el.style.top = "0";
              el.style.padding = "2px";
              el.style.backgroundColor = "#fff";
              el.style.borderRadius = "5px";
              el.style.overflow = "hidden";
              el.style.display = "flex";
              el.style.flexDirection = "column";
              el.style.alignItems = "stretch";
              el.style.justifyContent ="center";
              el.style.boxShadow = "0 4px 20px rgba(0,0,0,.5)";
            },
            createEl("div",
              function(el) {
                el.id = "no-more-404s-header";
                el.style.alignItems = "center";
                el.style.backgroundColor = "#0996f8";
                el.style.borderBottom = "1px solid #0675d3";
                el.style.borderRadius = "4px 4px 0 0";
                el.style.color = "#fff";
                el.style.display = "flex";
                el.style.fontSize = "24px";
                el.style.fontWeight = "700";
                el.style.height = "54px";
                el.style.justifyContent = "center";
                el.appendChild(document.createTextNode("Page missing? No worries!"));
              },
              createEl("button",
                function(el) {
                  el.style.position = "absolute";
                  el.style.display = "flex";
                  el.style.alignItems = "center";
                  el.style.justifyContent = "center";
                  el.style.transition = "background-color 150ms";
                  el.style.top = "12px";
                  el.style.right = "16px";
                  el.style.width = "22px";
                  el.style.height = "22px";
                  el.style.borderRadius = "3px";
                  el.style.boxSizing = "border-box";
                  el.style.padding = "0";
                  el.style.border = "none";
                  el.onclick = function() {
                    clearInterval(enforceBannerInterval);
                    document.getElementById("no-more-404s-message").style.display = "none";
                    bannerWasClosed = true;
                    sendTelemetry("dismissed");
                  };
                  el.onmouseenter = function() {
                    el.style.backgroundColor = "rgba(0,0,0,.1)";
                    el.style.boxShadow = "0 1px 0 0 rgba(0,0,0,.1) inset";
                  };
                  el.onmousedown = function() {
                    el.style.backgroundColor = "rgba(0,0,0,.2)";
                    el.style.boxShadow = "0 1px 0 0 rgba(0,0,0,.15) inset";
                  };
                  el.onmouseup = function() {
                    el.style.backgroundColor = "rgba(0,0,0,.1)";
                    el.style.boxShadow = "0 1px 0 0 rgba(0,0,0,.1) inset";
                  };
                  el.onmouseleave = function() {
                    el.style.backgroundColor = "transparent";
                    el.style.boxShadow = "";
                  };
                },
                createEl("img",
                  function(el) {
                    el.src = chrome.extension.getURL("images/close.svg");
                    el.alt = "close";
                    el.style.height = "16px";
                    el.style.transition = "background-color 100ms";
                    el.style.width = "16px";
                    el.style.margin = "0 auto";
                  }
                )
              )
            ),
            createEl("p", function(el) {
              el.appendChild(document.createTextNode("View a saved version courtesy of the Wayback Machine."));
              el.style.fontSize = "16px";
              el.style.margin = "20px 0 4px 0";
              el.style.textAlign = "center";
            }),
            createEl("img", function(el) {
              el.id = "no-more-404s-image";
              el.src = chrome.extension.getURL("images/car.gif");
              el.style.height = "300px";
              el.style.position = "relative";
              el.style.width = "420px";
            }),
            createEl("a", function(el) {
              el.id = "no-more-404s-message-link";
              el.href = wayback_url;
              el.style.alignItems = "center";
              el.style.backgroundColor = "#0996f8";
              el.style.border = "1px solid #0675d3";
              el.style.borderRadius = "3px";
              el.style.color = "#fff";
              el.style.display = "flex";
              el.style.fontSize = "19px";
              el.style.height = "52px";
              el.style.justifyContent = "center";
              el.style.margin = "20px";
              el.style.textDecoration = "none";
              el.appendChild(document.createTextNode("View saved version"));
              el.onmouseenter = function() {
                el.style.backgroundColor = "#0675d3";
                el.style.border = "1px solid #0568ba";
              };
              el.onmousedown = function() {
                el.style.backgroundColor = "#0568ba";
                el.style.border = "1px solid #0568ba";
              };
              el.onmouseup = function() {
                el.style.backgroundColor = "#0675d3";
                el.style.border = "1px solid #0568ba";
              };
              el.onmouseleave = function() {
                el.style.backgroundColor = "#0996f8";
                el.style.border = "1px solid #0675d3";
              };
              el.onclick = function(e) {
                archiveLinkWasClicked = true;

                // Work-around for myspace which hijacks the link
                if (window.location.hostname.indexOf("myspace.com") >= 0) {
                  sendTelemetry("viewed", function() {
                    setInterval(function() {
                      window.location.href = wayback_url;
                    }, 100);
                  });
                  e.preventDefault();
                  return false;
                } else {
                  sendTelemetry("viewed");
                }
              };
            })
          )
        )
      );
    // Focus the link for accessibility
    document.getElementById("no-more-404s-message-link").focus();

    // Transition element in from top of page
    setTimeout(function() {
      document.getElementById("no-more-404s-message").style.transform = "translate(0, 0%)";
    }, 100);

    bannerWasShown = true;
  }

  function checkIt(wayback_url) {
    // Some pages use javascript to update the dom so poll to ensure
    // the banner gets recreated if it is deleted.
    enforceBannerInterval = setInterval(function() {
      createBanner(wayback_url);
    }, 500);

    // Bind leave page for telemetry
    window.onunload = function() {
      if (bannerWasShown && !bannerWasClosed && !archiveLinkWasClicked) {
        sendTelemetry("ignored");
      }
    };
  }

  // Listen to message from background.js
  chrome.runtime.onMessage.addListener(
    function(request, sender, sendResponse) {
      if (request.type === "SHOW_BANNER") {
        if (request.wayback_url) {
          checkIt(request.wayback_url);
        }
      }
  });
})();


src/scripts/background.js
/*
 * License: AGPL-3
 * Copyright 2016, Internet Archive
 */
var VERSION = "1.8.1";

var excluded_urls = [
  "web.archive.org/web/",
  "localhost",
  "0.0.0.0",
  "127.0.0.1"
];

function isValidUrl(url) {
  for (var i = 0; i < excluded_urls.length; i++) {
    if (url.startsWith("http://" + excluded_urls[i]) || url.startsWith("https://" + excluded_urls[i])) {
      return false;
    }
  }
  return true;
}

/**
 * Header callback
 */
chrome.webRequest.onCompleted.addListener(function(details) {
  function tabIsReady(isIncognito) {
    var httpFailCodes = [404, 408, 410, 451, 500, 502, 503, 504,
                         509, 520, 521, 523, 524, 525, 526];
    if (isIncognito === false &&
        details.frameId === 0 &&
        httpFailCodes.indexOf(details.statusCode) >= 0 &&
        isValidUrl(details.url)) {
      wmAvailabilityCheck(details.url, function(wayback_url, url) {
        chrome.tabs.executeScript(details.tabId, {
          file: "scripts/client.js"
        }, function() {
          chrome.tabs.sendMessage(details.tabId, {
            type: "SHOW_BANNER",
            wayback_url: wayback_url
          });
        });
      }, function() {
        
      });
    }
  }
  chrome.tabs.get(details.tabId, function(tab) {
    tabIsReady(tab.incognito);
  });
}, {urls: ["<all_urls>"], types: ["main_frame"]});

/**
 * Checks Wayback Machine API for url snapshot
 */
function wmAvailabilityCheck(url, onsuccess, onfail) {
  var xhr = new XMLHttpRequest();
  var requestUrl = "https://archive.org/wayback/available";
  var requestParams = "url=" + encodeURI(url);
  xhr.open("POST", requestUrl, true);
  xhr.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
  xhr.setRequestHeader("User-Agent", navigator.userAgent + " Wayback_Machine_Firefox/" + VERSION);
  xhr.setRequestHeader("Wayback-Api-Version", 2);
  xhr.onload = function() {
    var response = JSON.parse(xhr.responseText);
    var wayback_url = getWaybackUrlFromResponse(response);
    if (wayback_url !== null) {
      onsuccess(wayback_url, url);
    } else if (onfail) {
      onfail();
    }
  };
  xhr.send(requestParams);
}

/**
 * @param response {object}
 * @return {string or null}
 */
function getWaybackUrlFromResponse(response) {
  if (response.results &&
      response.results[0] &&
      response.results[0].archived_snapshots &&
      response.results[0].archived_snapshots.closest &&
      response.results[0].archived_snapshots.closest.available &&
      response.results[0].archived_snapshots.closest.available === true &&
      response.results[0].archived_snapshots.closest.status.indexOf("2") === 0 &&
      isValidSnapshotUrl(response.results[0].archived_snapshots.closest.url)) {
    return makeHttps(response.results[0].archived_snapshots.closest.url);
  } else {
    return null;
  }
}

function makeHttps(url) {
  return url.replace(/^http:/, "https:");
}

/**
 * Makes sure response is a valid URL to prevent code injection
 * @param url {string}
 * @return {bool}
 */
function isValidSnapshotUrl(url) {
  return ((typeof url) === "string" &&
    (url.indexOf("http://") === 0 || url.indexOf("https://") === 0));
}


chrome.runtime.onMessage.addListener(function(message,sender,sendResponse){
  if(message.message=='openurl'){
    chrome.tabs.query({active: true, currentWindow: true}, function(tabs) {
      var tab = tabs[0];
      var page_url = tab.url;
      wayback_url = message.wayback_url;
      var pattern = /https:\/\/web\.archive\.org\/web\/(.+?)\//g;
      url = page_url.replace(pattern, "");
      open_url = wayback_url+encodeURI(url);
      if (message.method!='save') {
        wmAvailabilityCheck(url,function(){
          chrome.tabs.create({ url:  open_url});
        },function(){
          chrome.runtime.sendMessage({message:'urlnotfound'},function(response){
          });
        })
      } else {
        chrome.tabs.create({ url:  open_url});
      }
    });
  }
});

chrome.webRequest.onErrorOccurred.addListener(function(details) {
  function tabIsReady(isIncognito) {
    if(details.error == 'NS_ERROR_NET_ON_CONNECTING_TO'  || details.error == 'NS_ERROR_NET_ON_RESOLVED'){
      wmAvailabilityCheck(details.url, function(wayback_url, url) {
        chrome.tabs.update(details.tabId, {url: chrome.extension.getURL('dnserror.html')+"?url="+wayback_url});
      }, function() {
        
      });
    }
  }
  if(details.tabId >0 ){
    chrome.tabs.get(details.tabId, function(tab) {
      tabIsReady(tab.incognito);
    });  
  }

  
}, {urls: ["<all_urls>"], types: ["main_frame"]});